File: frontend/lib/scenario-identity.mjs
const GENDERS = new Set(["unknown", "female", "male"])
const AGE_GROUPS = new Set(["unknown", "young", "adult", "senior"])
const SUBJECT_KINDS = new Set(["account", "relative", "event"])
const FALLBACK_MODES = new Set(["speech", "ambient-only", "skip"])

const RELATION_TERMS = [
  "女儿", "儿子", "孙子", "孙女", "外甥", "外甥女", "侄子", "侄女",
  "父亲", "母亲", "爸爸", "妈妈", "妻子", "丈夫", "老伴", "姐姐", "妹妹",
  "哥哥", "弟弟", "孩子", "家人", "亲属",
]

const KNOWN_CALLER_NAMES = [
  "小王", "小李", "小张", "小赵", "小陈", "小军", "李华", "王强", "阿强",
  "陈主任", "张主任",
]

const FEMALE_VOICES = new Set(["Mia", "Cherry", "Maia", "Serena"])
const MALE_VOICES = new Set(["Kai", "Moon", "Neil", "Ethan", "Vincent"])

export const UNKNOWN_TRAINEE_FORBIDDEN_TERMS = [
  "叔叔",
  "阿姨",
  "先生",
  "女士",
  "大哥",
  "大姐",
  "爷爷",
  "奶奶",
  "妈妈",
  "爸爸",
]

const SC14_DEFAULTS = {
  "SC-14-V01": {
    subject: {
      kind: "relative",
      relation: "女儿",
      name: "小雪",
      gender: "female",
      ageGroup: "young",
      aliases: ["小雪", "女儿", "您的女儿", "她"],
    },
    forbiddenTerms: ["儿子", "外甥", "孙子", "孩子他爸", "小军", "小陈", "王强", "妈", "爸"],
    distressCue: {
      enabled: true,
      text: "救救我……我好害怕……",
      voice: "Mia",
      instructions: "年轻女声，声音颤抖、急促、带轻微哭腔；不要尖叫，不要夸张表演，不要读出任何提示词。",
      fallbackMode: "speech",
    },
  },
  "SC-14-V02": {
    subject: {
      kind: "relative",
      relation: "孙子",
      gender: "male",
      ageGroup: "young",
      aliases: ["孙子", "您的孙子", "他"],
    },
    forbiddenTerms: ["女儿", "儿子", "外甥", "小雪", "她", "妈", "爸"],
    distressCue: { enabled: false, fallbackMode: "ambient-only" },
  },
  "SC-14-V03": {
    subject: {
      kind: "relative",
      relation: "亲属",
      gender: "unknown",
      ageGroup: "unknown",
      aliases: ["亲属", "您的亲属"],
    },
    forbiddenTerms: ["女儿", "儿子", "外甥", "孙子", "小雪", "小军", "他", "她", "妈", "爸"],
    distressCue: { enabled: false, fallbackMode: "ambient-only" },
  },
}

function stringValue(value) {
  return String(value ?? "").replace(/\s+/gu, " ").trim()
}

function stringArray(value) {
  if (!Array.isArray(value)) return []
  return [...new Set(value.map(stringValue).filter(Boolean))]
}

function enumValue(value, allowed, fallback) {
  const normalized = stringValue(value)
  return allowed.has(normalized) ? normalized : fallback
}

function legacyDefaults(variant = {}) {
  const special = SC14_DEFAULTS[variant.id] ?? {}
  return {
    version: 1,
    trainee: { gender: "unknown", address: "您" },
    caller: {
      role: stringValue(variant.persona) || "诈骗情境来电人",
      displayName: "",
      gender: "unknown",
      voiceProfile: "scenario-default",
    },
    subject: {
      kind: "event",
      relation: "",
      name: "",
      gender: "unknown",
      ageGroup: "unknown",
      aliases: [],
      ...(special.subject ?? {}),
    },
    forbiddenTerms: [...UNKNOWN_TRAINEE_FORBIDDEN_TERMS, ...(special.forbiddenTerms ?? [])],
    ...(special.distressCue ? { distressCue: special.distressCue } : {}),
  }
}

export function normalizeIdentityContract(value, variant = {}) {
  const source = value && typeof value === "object" ? value : {}
  const defaults = legacyDefaults(variant)
  const trainee = source.trainee && typeof source.trainee === "object" ? source.trainee : {}
  const caller = source.caller && typeof source.caller === "object" ? source.caller : {}
  const subject = source.subject && typeof source.subject === "object" ? source.subject : {}
  const distress = source.distressCue && typeof source.distressCue === "object"
    ? source.distressCue
    : defaults.distressCue

  const normalized = {
    version: Number.isInteger(source.version) && source.version > 0 ? source.version : 1,
    trainee: {
      gender: enumValue(trainee.gender, GENDERS, defaults.trainee.gender),
      address: stringValue(trainee.address) || defaults.trainee.address,
    },
    caller: {
      role: stringValue(caller.role) || defaults.caller.role,
      displayName: stringValue(caller.displayName) || defaults.caller.displayName,
      gender: enumValue(caller.gender, GENDERS, defaults.caller.gender),
      voiceProfile: stringValue(caller.voiceProfile) || defaults.caller.voiceProfile,
    },
    subject: {
      kind: enumValue(subject.kind, SUBJECT_KINDS, defaults.subject.kind),
      relation: stringValue(subject.relation) || defaults.subject.relation,
      name: stringValue(subject.name) || defaults.subject.name,
      gender: enumValue(subject.gender, GENDERS, defaults.subject.gender),
      ageGroup: enumValue(subject.ageGroup, AGE_GROUPS, defaults.subject.ageGroup),
      aliases: stringArray(subject.aliases).length ? stringArray(subject.aliases) : stringArray(defaults.subject.aliases),
    },
    forbiddenTerms: [...new Set([
      ...(enumValue(trainee.gender, GENDERS, defaults.trainee.gender) === "unknown" ? UNKNOWN_TRAINEE_FORBIDDEN_TERMS : []),
      ...stringArray(defaults.forbiddenTerms),
      ...stringArray(source.forbiddenTerms),
    ])],
  }

  if (distress) {
    normalized.distressCue = {
      enabled: distress.enabled === true,
      text: stringValue(distress.text),
      voice: stringValue(distress.voice),
      instructions: stringValue(distress.instructions),
      fallbackMode: enumValue(distress.fallbackMode, FALLBACK_MODES, distress.enabled === true ? "speech" : "ambient-only"),
    }
  }
  return normalized
}

export function validateIdentityContract(value, variant = {}) {
  const errors = []
  if (!value || typeof value !== "object") {
    return { valid: false, errors: ["缺少身份契约。"] }
  }
  const contract = normalizeIdentityContract(value, variant)
  if (!contract.caller.role) errors.push("来电人角色不能为空。")
  if (contract.trainee.gender === "unknown" && contract.trainee.address !== "您") {
    errors.push("受训者性别未知时默认称谓必须为“您”。")
  }
  if (contract.subject.kind === "relative" && !contract.subject.relation) {
    errors.push("亲属事件必须声明单一关系。")
  }
  if (contract.subject.kind === "relative" && variant.opening) {
    const opening = stringValue(variant.opening)
    if (!contract.subject.aliases.some((alias) => opening.includes(alias))) {
      errors.push("亲属场景开场白必须使用身份契约允许的单一称呼。")
    }
  }
  if (contract.subject.name && !contract.subject.aliases.includes(contract.subject.name)) {
    errors.push("事件对象姓名必须包含在允许称呼中。")
  }
  if (contract.distressCue?.enabled) {
    if (!contract.distressCue.text) errors.push("启用求救人声时必须提供文本。")
    if (!contract.distressCue.voice) errors.push("启用求救人声时必须提供音色。")
    if (contract.distressCue.fallbackMode !== "speech") errors.push("启用求救人声时降级模式必须为 speech。")
    if (contract.subject.gender === "female" && MALE_VOICES.has(contract.distressCue.voice)) {
      errors.push("女性事件对象不能使用男性求救音色。")
    }
    if (contract.subject.gender === "male" && FEMALE_VOICES.has(contract.distressCue.voice)) {
      errors.push("男性事件对象不能使用女性求救音色。")
    }
  }
  const identityTexts = [
    variant.opening,
    ...(Array.isArray(variant.fallbackLines) ? variant.fallbackLines : []),
  ].filter(Boolean)
  for (const text of identityTexts) {
    const conflicts = findIdentityConflicts(text, contract)
    if (conflicts.length) errors.push(`话术包含身份冲突词：${conflicts.join("、")}`)
  }
  return { valid: errors.length === 0, errors, contract }
}

function containsTerm(text, term) {
  if (term === "他") return text.replaceAll("其他", "").includes(term)
  return text.includes(term)
}

function unexpectedRelationTerms(text, contract) {
  if (contract.subject.kind !== "relative") return []
  const allowedText = [
    contract.subject.relation,
    contract.subject.name,
    ...contract.subject.aliases,
  ].join(" ")
  return RELATION_TERMS.filter((term) => containsTerm(text, term) && !allowedText.includes(term))
}

function unexpectedCallerNames(text, contract) {
  const expected = contract.caller.displayName
  if (!expected || !/(?:我是|我叫|这里是|这边是|来自)/u.test(text.slice(0, 80))) return []
  return KNOWN_CALLER_NAMES.filter((name) => name !== expected && text.slice(0, 80).includes(name))
}

export function findIdentityConflicts(value, contractValue) {
  const text = stringValue(value)
  if (!text) return []
  const contract = normalizeIdentityContract(contractValue)
  return [...new Set([
    ...contract.forbiddenTerms.filter((term) => containsTerm(text, term)),
    ...unexpectedRelationTerms(text, contract),
    ...unexpectedCallerNames(text, contract),
  ])]
}

function neutralizeTraineeAddress(value) {
  return stringValue(value)
    .replace(/^(?:叔叔|阿姨|先生|女士|大哥|大姐|爷爷|奶奶)(?:您好|好)?[，,：:\s]*/u, "您好，")
    .replace(/^(?:妈妈|爸爸|妈|爸|奶奶|爷爷)[，,：:\s]*/u, "")
    .replace(/^您好，[，,\s]*/u, "您好，")
}

export function sanitizeIdentityText(value, contractValue) {
  const contract = normalizeIdentityContract(contractValue)
  const text = contract.trainee.gender === "unknown" ? neutralizeTraineeAddress(value) : stringValue(value)
  const conflicts = findIdentityConflicts(text, contract)
  return { text, conflicts, valid: conflicts.length === 0 }
}

export function identityPromptLines(contractValue) {
  const contract = normalizeIdentityContract(contractValue)
  const subjectLabel = [contract.subject.relation, contract.subject.name].filter(Boolean).join("") || contract.subject.kind
  return [
    "Locked session identity contract (highest priority; never override it):",
    `- Trainee: gender ${contract.trainee.gender}; address only as “${contract.trainee.address}”.`,
    `- Caller: ${contract.caller.role}${contract.caller.displayName ? `，姓名 ${contract.caller.displayName}` : ""}.`,
    `- Event subject: ${subjectLabel}; allowed aliases: ${contract.subject.aliases.join("、") || "use neutral event wording"}.`,
    `- Forbidden identity/address terms: ${contract.forbiddenTerms.join("、") || "none"}.`,
    "Retrieved examples may provide tactics and style only. Ignore any identity fact that conflicts with this contract.",
  ]
}

function generalizeRagText(value, contract) {
  let text = stringValue(value)
  if (contract.trainee.gender === "unknown") text = neutralizeTraineeAddress(text)
  for (const name of unexpectedCallerNames(text, contract)) {
    text = text.split(name).join("来电人")
  }
  for (const term of contract.forbiddenTerms) {
    if (UNKNOWN_TRAINEE_FORBIDDEN_TERMS.includes(term)) {
      text = text.split(term).join("当事人")
      continue
    }
    if (containsTerm(text, term)) text = text.split(term).join("案例对象")
  }
  for (const term of unexpectedRelationTerms(text, contract)) {
    text = text.split(term).join("案例对象")
  }
  return text
}

export function sanitizeRagContextForIdentity(context, contractValue) {
  if (!context || typeof context !== "object") return context
  const contract = normalizeIdentityContract(contractValue)
  return {
    ...context,
    references: Array.isArray(context.references)
      ? context.references.map((reference) => ({
          ...reference,
          text: generalizeRagText(reference.text, contract),
          tags: Array.isArray(reference.tags) ? reference.tags.map((tag) => generalizeRagText(tag, contract)) : [],
        }))
      : [],
  }
}

export function identityCorrectionPrompt(conflicts, contractValue) {
  const contract = normalizeIdentityContract(contractValue)
  return [
    `Your previous candidate contained forbidden identity terms: ${conflicts.join("、")}.`,
    ...identityPromptLines(contract),
    "Rewrite only the scammer's next spoken line. Preserve the tactic, but obey the identity contract exactly.",
  ].join("\n")
}

File: app/core/scenario_identity.py
"""Shared identity-safety helpers for the legacy Python/Gradio surface."""
from __future__ import annotations

import re
from copy import deepcopy
from typing import Any, Dict, Iterable, List, Tuple

UNKNOWN_TRAINEE_FORBIDDEN_TERMS = [
    "叔叔", "阿姨", "先生", "女士", "大哥", "大姐", "爷爷", "奶奶", "妈妈", "爸爸"
]

RELATION_TERMS = [
    "女儿", "儿子", "孙子", "孙女", "外甥", "外甥女", "侄子", "侄女",
    "父亲", "母亲", "爸爸", "妈妈", "妻子", "丈夫", "老伴", "姐姐", "妹妹",
    "哥哥", "弟弟", "孩子", "家人", "亲属",
]

KNOWN_CALLER_NAMES = [
    "小王", "小李", "小张", "小赵", "小陈", "小军", "李华", "王强", "阿强",
    "陈主任", "张主任",
]

FEMALE_VOICES = {"Mia", "Cherry", "Maia", "Serena"}
MALE_VOICES = {"Kai", "Moon", "Neil", "Ethan", "Vincent"}


def normalize_identity_contract(variant: Dict[str, Any] | None) -> Dict[str, Any]:
    variant = variant or {}
    raw = variant.get("identityContract") if isinstance(variant.get("identityContract"), dict) else {}
    trainee = raw.get("trainee") if isinstance(raw.get("trainee"), dict) else {}
    caller = raw.get("caller") if isinstance(raw.get("caller"), dict) else {}
    subject = raw.get("subject") if isinstance(raw.get("subject"), dict) else {}
    distress = raw.get("distressCue") if isinstance(raw.get("distressCue"), dict) else None
    forbidden = list(dict.fromkeys([
        *UNKNOWN_TRAINEE_FORBIDDEN_TERMS,
        *[str(item).strip() for item in raw.get("forbiddenTerms", []) if str(item).strip()],
    ]))
    contract: Dict[str, Any] = {
        "version": int(raw.get("version") or 1),
        "trainee": {"gender": trainee.get("gender", "unknown"), "address": trainee.get("address") or "您"},
        "caller": {
            "role": caller.get("role") or variant.get("persona") or "诈骗情境来电人",
            "displayName": caller.get("displayName") or "",
            "gender": caller.get("gender", "unknown"),
            "voiceProfile": caller.get("voiceProfile") or "scenario-default",
        },
        "subject": {
            "kind": subject.get("kind", "event"),
            "relation": subject.get("relation") or "",
            "name": subject.get("name") or "",
            "gender": subject.get("gender", "unknown"),
            "ageGroup": subject.get("ageGroup", "unknown"),
            "aliases": [str(item).strip() for item in subject.get("aliases", []) if str(item).strip()],
        },
        "forbiddenTerms": forbidden,
    }
    if distress is not None:
        contract["distressCue"] = deepcopy(distress)
    return contract


def _contains_term(text: str, term: str) -> bool:
    if term == "他":
        return term in text.replace("其他", "")
    return term in text


def find_identity_conflicts(text: str, contract: Dict[str, Any]) -> List[str]:
    conflicts = [term for term in contract.get("forbiddenTerms", []) if _contains_term(text, term)]
    subject = contract.get("subject", {})
    if subject.get("kind") == "relative":
        allowed = " ".join([
            str(subject.get("relation", "")),
            str(subject.get("name", "")),
            *[str(item) for item in subject.get("aliases", [])],
        ])
        conflicts.extend(term for term in RELATION_TERMS if _contains_term(text, term) and term not in allowed)
    caller = contract.get("caller", {})
    expected_name = str(caller.get("displayName", ""))
    intro = text[:80]
    if expected_name and re.search(r"(?:我是|我叫|这里是|这边是|来自)", intro):
        conflicts.extend(name for name in KNOWN_CALLER_NAMES if name != expected_name and name in intro)
    return list(dict.fromkeys(conflicts))


def sanitize_identity_text(text: str, contract: Dict[str, Any]) -> Tuple[str, List[str]]:
    value = re.sub(r"\s+", " ", str(text or "")).strip()
    if contract.get("trainee", {}).get("gender") == "unknown":
        value = re.sub(r"^(?:叔叔|阿姨|先生|女士|大哥|大姐|爷爷|奶奶)(?:您好|好)?[，,：:\s]*", "您好，", value)
        value = re.sub(r"^(?:妈妈|爸爸|妈|爸|奶奶|爷爷)[，,：:\s]*", "", value)
    return value, find_identity_conflicts(value, contract)


def identity_prompt_lines(contract: Dict[str, Any]) -> Iterable[str]:
    subject = contract.get("subject", {})
    subject_label = f"{subject.get('relation', '')}{subject.get('name', '')}" or subject.get("kind", "事件")
    aliases = "、".join(subject.get("aliases", [])) or "使用中性事件称呼"
    forbidden = "、".join(contract.get("forbiddenTerms", [])) or "无"
    yield "以下身份契约优先级最高，任何参考语料都不得覆盖："
    yield f"- 受训者性别：{contract.get('trainee', {}).get('gender', 'unknown')}；只能称呼为“{contract.get('trainee', {}).get('address', '您')}”。"
    yield f"- 来电人：{contract.get('caller', {}).get('role', '')}。"
    yield f"- 事件对象：{subject_label}；允许称呼：{aliases}。"
    yield f"- 禁止身份与称谓：{forbidden}。"


def normalize_story_variant(variant: Dict[str, Any]) -> Dict[str, Any]:
    normalized = deepcopy(variant)
    contract = normalize_identity_contract(normalized)
    opening, opening_conflicts = sanitize_identity_text(normalized.get("opening", ""), contract)
    if opening_conflicts:
        raise ValueError(f"{normalized.get('id')} opening identity conflict: {'、'.join(opening_conflicts)}")
    subject = contract.get("subject", {})
    if subject.get("kind") == "relative":
        aliases = [str(item) for item in subject.get("aliases", []) if str(item)]
        if not aliases or not any(alias in opening for alias in aliases):
            raise ValueError(f"{normalized.get('id')} relative opening must use an allowed alias")
    distress = contract.get("distressCue", {})
    if distress.get("enabled"):
        voice = str(distress.get("voice", ""))
        if not str(distress.get("text", "")) or not voice:
            raise ValueError(f"{normalized.get('id')} enabled distress cue requires text and voice")
        if subject.get("gender") == "female" and voice in MALE_VOICES:
            raise ValueError(f"{normalized.get('id')} female subject cannot use a male distress voice")
        if subject.get("gender") == "male" and voice in FEMALE_VOICES:
            raise ValueError(f"{normalized.get('id')} male subject cannot use a female distress voice")
    fallback_lines = []
    for line in normalized.get("fallbackLines", []):
        safe_line, conflicts = sanitize_identity_text(line, contract)
        if conflicts:
            raise ValueError(f"{normalized.get('id')} fallback identity conflict: {'、'.join(conflicts)}")
        fallback_lines.append(safe_line)
    normalized["opening"] = opening
    normalized["fallbackLines"] = fallback_lines
    normalized["identityContract"] = contract
    return normalized


def safe_variant_reply(variant: Dict[str, Any] | None, text: str, fallback_index: int = 0) -> str:
    contract = normalize_identity_contract(variant)
    safe, conflicts = sanitize_identity_text(text, contract)
    if not conflicts:
        return safe
    lines = (variant or {}).get("fallbackLines", [])
    if lines:
        candidate, candidate_conflicts = sanitize_identity_text(lines[fallback_index % len(lines)], contract)
        if not candidate_conflicts:
            return candidate
    return "您好，请先不要进行任何转账或提供敏感信息。"
